articles/per_package_documentation.md

```diff
@@ -0,0 +1,115 @@
+
+**Package documentation must be maintainable in its repository without going through a third party**
+
+Maintainers and contributors will only have to work on their package's documentation within its repository.
+The details of building and hosting will not be a concern of package maintainers.
+
+Despite being hosted alongside the generic documentation, working on a package's documentation will not require any work on the repositories of the generic documentation (currently `ros2/ros2_documentation`) or the site repository (currently `ros-infrastructure/rosindex`).
+
+**The system must support C++ and Python API docs generation**
+
+The system will automatically extract API docs for these languages from the source code in package repositories and build the output on the docs site.
+
+**The system must allow package documentation to be versioned per ROS 2 distribution**
+
+The docs site and buildfarm will allow documentation for the latest version of a package corresponding to each ROS 2 distribution the docs site supports versioning for.
+The package repository can still maintain it's own docs for previous versions.
+
+While multiple package versions per distribution will not initially be supported, the URL structure will support the addition of this feature in the future (mentioned in the Secondary requirements below).
+
+**The buildfarm must automatically build package documentation**
+
+Changes to the documentation in a package repository shouldn't require the maintainer to manually trigger a build.
+The changes should be automatically built and uploaded to the site.
+
+**The system must automatically generate content for a package so it's listed on docs.ros.org even if the package maintainer does not explicitly set up anything in the package repository**
+
+Maintainers can explicitly opt out of being listed, but otherwise by default some package info will be listed for every available package.
+
+**The systems must support cross-referencing between packages**
+
+A stable linking process will be in place to cross-reference between package docs despite their originating from separate repositories.
+
+**The system must prevent the possibility of package file names (created by maintainers) colliding with auto-generated content (from the system's side)**
+
+When writing package documentation, maintainers should not have to concern themselves with auto-generated content, like the current `/changelogs` and `/symbols` directories in the current `docs.ros.org` API docs structure for ROS 1.
+
+### Secondary requirements
+
+**Package maintainers should have the ability to include free form documentation, like tutorials, design docs, etc., alongside generated API docs on docs.ros.org**
+
+**The system should provide a standardized interface for generating an index page/landing page for packages, consistent across ROS 2**
+
+**The system should provide a standardized method for generating package documentation locally, including build error alerts**
```

Comment:
I feel like this should be a primary requirement - otherwise the round-trip time required to create properly formatted documentation is going to be quite long (merge a change into the distro branch, wait for the build farm to catch the problem, then update and repeat) - and the net result will likely be many people give up and have either broken or very little documentation.

Also, if there is a method for building package documentation locally, it can also be added to the package CI (I imagine that tools like industrial_ci would quickly integrate this) - otherwise the effort for maintainers to check over documentation changes is going to be quite high.

Reply:
While I agree that being able to build the documentation locally is important, we made it a secondary requirement because it is something we can add later.  That is, it's the difference between having a system at all, and having a system that is nice to use.

Reply:
I agree that this should be something that gets implemented first just because it will greatly increase ease of use 

But our definition for secondary requirements is that they're important but "not necessary to roll out the first stage of implementation", which this does fall under. We need to ensure that we can move quickly through the initial implementation and not get tied up in anything that isn't an _absolutely_ necessity. So, we will definitely add this feature and all the others under "secondary requirements", just not at the first roll-out.

Reply:
I think it's OK to not have it straight away provided we get it really soon after. GitHub Actions provides a good example of why having to go via the online version makes using the tool _extremely_ painful.

Reply:
> GitHub Actions provides a good example of why having to go via the online version makes using the tool *extremely* painful.

Off-topic, but: yes (although: [nektos/act](https://github.com/nektos/act)).

Reply:
> Is it correct to state that the secondary requirements are still absolute requirements on how the system is architected?

Yes, exactly.  That's a great way to put it.